### PR TITLE
[Feature] Support TimeSformer

### PR DESCRIPTION
**Pull request by congee524**
- [x] Attention Code
- [x] TimeSformer Backbone Code
- [x] TimeSformer Head
- [x] Inference
- [x] Train
- [x] CheckPoint
- [x] Refine Code
- [x] Add Docstring
- [x] Unittest
- [x] README

### CODE REVIEW

**Review by kennymckormick — COMMENTED** [review]
The current min version of mmcv is 1.3.1, is it sufficient for this PR?

**Comment by kennymckormick** [comment]
> The current min version of mmcv is 1.3.1, is it sufficient for this PR?

Oh, seems not. You need to update the version requirement as well as the CI. Besides, you can add some unittest to improve the CodeCov

File: configs/recognition/timesformer/timesformer_divST_8x32x1_15e_kinetics400_rgb.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        embed_dims=768,
+        in_channels=3,
+        dropout_ratio=0.,
+        transformer_layers=None,
+        attention_type='divided_space_time',
+        norm_cfg=dict(type='LN', eps=1e-6)),
+    cls_head=dict(type='TimeSformerHead', num_classes=400, in_channels=768),
+    # model training and testing settings
+    train_cfg=None,
+    test_cfg=dict(average_clips='score'))
```
[the remark this answers is not in the document]
Comment (congee524) [reply configs/recognition/timesformer/timesformer_divST_8x32x1_15e_kinetics400_rgb.py:22]: My fault. mistake `softmax scores` for `scores`

File: mmaction/models/backbones/timesformer.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        num_heads (int): Number of parallel attention heads in
+            TransformerCoder. Defaults to 12.
+        num_transformer_layers (int): Number of transformer layers. Defaults to
+            12.
+        in_channels (int): Channel num of input features. Defaults to 3.
+        dropout_ratio (float): Probability of dropout layer. Defaults to 0..
+        transformer_layers (list[obj:`mmcv.ConfigDict`] |
+            obj:`mmcv.ConfigDict`): Config of transformerlayer in
+            TransformerCoder. If it is obj:`mmcv.ConfigDict`, it would be
+            repeated `num_transformer_layers` times to a
```
[the remark this answers is not in the document]
Comment (congee524) [reply mmaction/models/backbones/timesformer.py:84]: > > The current min version of mmcv is 1.3.1, is it sufficient for this PR?
> 
> Oh, seems not. You need to update the version requirement as well as the CI. Besides, you can add some unittest to improve the CodeCov

OK. The work may need to wait until mmcv 1.3.5 release.

**Comment by congee524** [comment]
> > The current min version of mmcv is 1.3.1, is it sufficient for this PR?
> 
> Oh, seems not. You need to update the version requirement as well as the CI. Besides, you can add some unittest to improve the CodeCov

OK. The work may need to wait until mmcv 1.3.5 release.

**Comment by kennymckormick** [comment]
> > > The current min version of mmcv is 1.3.1, is it sufficient for this PR?
> > 
> > 
> > Oh, seems not. You need to update the version requirement as well as the CI. Besides, you can add some unittest to improve the CodeCov
> 
> OK. The work may need to wait until mmcv 1.3.5 release.

You can ping me after this PR is ready.

File: mmaction/models/common/transformer.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                 dropout_layer=dict(type='DropPath', drop_prob=0.1),
+                 norm_cfg=dict(type='LN'),
+                 init_cfg=None,
+                 **kwargs):
+        super().__init__(init_cfg)
+        self.embed_dims = embed_dims
+        self.num_heads = num_heads
+        self.num_frames = num_frames
+        self.norm = build_norm_layer(norm_cfg, self.embed_dims)[1]
+        self.attn = nn.MultiheadAttention(embed_dims, num_heads, attn_drop,
```
Comment (kennymckormick) [inline mmaction/models/common/transformer.py:46]: Perhaps you need to use MultiheadAttention from mmcv since it takes the arg `batch_first`

File: mmaction/models/common/transformer.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
```
Comment (kennymckormick) [inline mmaction/models/common/transformer.py:121]: Same here

File: mmaction/models/common/transformer.py (hunk 3 above)
Comment (shvdiwnkozbw) [reply mmaction/models/common/transformer.py:46]: Yes, it should be turned into mmcv MultiheadAttention, pytorch 1.9 includes arg but previous versions not.

**Comment by shvdiwnkozbw** [comment]
It seems MMCV minimum version 1.3.6 is required.

**Comment by kennymckormick** [comment]
@shvdiwnkozbw The old checkpoints also need to be updated. You can send me the converted checkpoint.

**Comment by shvdiwnkozbw** [comment]
> @shvdiwnkozbw The old checkpoints also need to be updated. You can send me the converted checkpoint.

This is the link for converted checkpoint
https://drive.google.com/drive/folders/1E3uwhfDZlArzI8wsvGGD26EoNbrxsUp7?usp=sharing

**Comment by kennymckormick** [comment]
> > @shvdiwnkozbw The old checkpoints also need to be updated. You can send me the converted checkpoint.
> 
> This is the link for converted checkpoint
> https://drive.google.com/drive/folders/1E3uwhfDZlArzI8wsvGGD26EoNbrxsUp7?usp=sharing

Have you tested the converted checkpoints? It seems that the testing accuracy doesn't align with the reported numbers, e.g., I got 61.8% Top-1 when I test the divST model.

**Comment by shvdiwnkozbw** [comment]
> Have you tested the converted checkpoints? It seems that the testing accuracy doesn't align with the reported numbers, e.g., I got 61.8% Top-1 when I test the divST model (jointST failed).

I have not tested the accuracy since my account has been cancelled, are there any missing parameters when loading the converted checkpoint?

**Comment by kennymckormick** [comment]
> > Have you tested the converted checkpoints? It seems that the testing accuracy doesn't align with the reported numbers, e.g., I got 61.8% Top-1 when I test the divST model (jointST failed).
> 
> I have not tested the accuracy since my account has been cancelled, are there any missing parameters when loading the converted checkpoint?

For divST, the parameter names aligned perfectly but the accuracy is about 10% inferior. For jointST, the checkpoint can not be loaded properly.

**Review by kennymckormick — APPROVED** [review]
